internal/k8sCommon/k8sclient/endpointslicewatcher.go

```
@@ -0,0 +1,297 @@
+
+	// For bookkeeping, so we can remove old mappings upon EndpointSlice deletion
+	sliceToKeysMap sync.Map // map[sliceUID string] -> []string of keys we inserted, which are "ip" or "ip:port"
+	deleter        Deleter
+}
+
+// PodMetadata holds {"workload", "namespace", "node"}
+type PodMetadata struct {
+	Workload  string
+	Namespace string
+	Node      string
+}
+
+// kvPair holds one mapping from key -> value. The isService flag
+// indicates whether this key is for a Service or for an IP/IP:port.
+type kvPair struct {
+	key   string      // key: "ip" or "ip:port"
+	value PodMetadata // value: {"workload", "namespace", "node"}
+}
+
+// NewEndpointSliceWatcher creates an EndpointSlice watcher
+func NewEndpointSliceWatcher(
+	logger *zap.Logger,
+	factory informers.SharedInformerFactory,
+	deleter Deleter,
+) *EndpointSliceWatcher {
+
+	esInformer := factory.Discovery().V1().EndpointSlices().Informer()
+	err := esInformer.SetTransform(minimizeEndpointSlice)
+	if err != nil {
+		logger.Error("failed to minimize Service objects", zap.Error(err))
+	}
+
+	return &EndpointSliceWatcher{
+		logger:          logger,
+		informer:        esInformer,
+		IPToPodMetadata: &sync.Map{},
+		deleter:         deleter,
+	}
+}
+
+// run starts the endpointSliceWatcher.
+func (w *EndpointSliceWatcher) Run(stopCh chan struct{}) {
+	w.informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			w.handleSliceAdd(obj)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			w.handleSliceUpdate(oldObj, newObj)
+		},
+		DeleteFunc: func(obj interface{}) {
+			w.handleSliceDelete(obj)
+		},
+	})
+	go w.informer.Run(stopCh)
+}
+
+func (w *EndpointSliceWatcher) WaitForCacheSync(stopCh chan struct{}) {
+	if !cache.WaitForNamedCacheSync("endpointSliceWatcher", stopCh, w.informer.HasSynced) {
+		w.logger.Error("timed out waiting for endpointSliceWatcher cache to sync")
```

should we return an error here? 

What happens if the cache fails to sync?

---

If the cache fails to sync, then we will just have outdated and potentially missing metadata. It won't break anything, we just won't get an accurate entity. Therefore, we error.